File: LinkedList-Stack-Queue/Q86.py
# https://leetcode.com/problems/partition-list/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def partition(self, head: ListNode, x: int) -> ListNode:
        ptrLess = less = ListNode()
        ptrOther = other = ListNode()

        while head:
            if head.val < x:
                ptrLess.next = head
                ptrLess = ptrLess.next
            else:
                ptrOther.next = head
                ptrOther = ptrOther.next
            head = head.next

        if ptrLess:
            ptrLess.next = other.next
            if ptrOther:
                ptrOther.next = None
            return less.next
        else:
            return other.next

File: LinkedList-Stack-Queue/Q86.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} x
 * @return {ListNode}
 */
var partition = function(head, x) {
  let LT = new ListNode(), ptrLT = LT;
  let GTE = new ListNode(), ptrGTE = GTE;
  
  while (head) {
    if (head.val < x) {
      ptrLT.next = new ListNode(head.val);
      ptrLT = ptrLT.next;
    } else {
      ptrGTE.next = new ListNode(head.val);
      ptrGTE = ptrGTE.next;
    }
    head = head.next;
  }
  
  ptrLT.next = GTE.next;
  return LT.next;
};
